pageserver/src/keyspace.rs

```diff
@@ -12,10 +12,22 @@ pub struct KeySpace {
     pub ranges: Vec<Range<Key>>,
 }
 
+//
+// See comment to partition() method below. This constant is limited each layer to contain 16k relation.
+// 16k - is first user Oid, so it separates system and user relations.
+//
+#[allow(clippy::unusual_byte_groupings)]
+const MAX_LAYER_KEY_RANGE: i128 = 0x00_00000000_00000000_00004000_00_00000000i128;
+
 impl KeySpace {
     ///
     /// Partition a key space into roughly chunks of roughly 'target_size' bytes
-    /// in each partition.
+    /// in each partition. But also limit key range of each partition to avoid mixing of rel/nonrel entries
+    /// and system/catalog relations. With current key encoding, frequently updated objects belong to opposite ends of key
+    /// dimension range. It means that layers generated after compaction are used to cover all database space.
+    /// Which cause image layer generation for the whole database, leading to huge rite amplification.
+    /// Catalog tables (like pg_class) are also used to be updated frequently (for example with estimated value of relation rows/size).
+    /// Even if we have on append only table, still generated delta layers will cover all this table, despite to the fact that only tail is updated.
     ///
     pub fn partition(&self, target_size: u64) -> KeyPartitioning {
```

Comment:
It would be a good idea to have a `#[test]` example for this fn, even though the implementation is not too complicated yet. Though, I cannot see an easy example right away.